File: javascript/beer-song/beer-song.js
var BeerSong = function(){};

BeerSong.prototype.verse = function(num) {
        if (num > 2) {
            return (num + " bottles of beer on the wall, " + num + " bottles of beer.\nTake one down and pass it around, " + (num - 1) + " bottles of beer on the wall.\n")
        } 
        if (num === 2 ) {
            return (num + " bottles of beer on the wall, " + num + " bottles of beer.\nTake one down and pass it around, " + (num - 1) + " bottle of beer on the wall.\n")
        } 
        if (num === 1) {
            return (num + " bottle of beer on the wall, " + num + " bottle of beer.\nTake it down and pass it around, no more bottles of beer on the wall.\n")
        } 
        if (num == 0) {
            return 'No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n'
    }
}

BeerSong.prototype.sing = function(bottles, rounds) {
    result = ""
    var self = this;
    rounds = (typeof rounds !== 'undefined') ?  rounds : 1;
    limit = (rounds !== 1) ? bottles - rounds : bottles    
    for (i=0; i<(limit + 1); i++) {
        result += self.verse.call(self, bottles) + "\n"
        bottles--
    }
    return result.slice(0,-1)
}

module.exports = BeerSong;

File: ruby/beer-song/beer_song.rb
class BeerSong
  STANDARD = <<~TEXT
    %<number>s %<bottle_or_bottles_before>s of beer on the wall, %<number>s %<bottle_or_bottles_before>s of beer.
    Take %<one_or_it>s down and pass it around, %<number_less_one>s %<bottle_or_bottles_after>s of beer on the wall.
    TEXT

  ZERO_BOTTLES = <<~TEXT
    No more bottles of beer on the wall, no more bottles of beer.
    Go to the store and buy some more, 99 bottles of beer on the wall.
    TEXT

  def verse(number_of_bottles)
    if (1..99).include?(number_of_bottles)
      text = STANDARD
    elsif number_of_bottles.zero?
      text = ZERO_BOTTLES
    end
    text % { number: number_of_bottles,
             number_less_one: number_less_one(number_of_bottles),
             bottle_or_bottles_before: bottle_or_bottles(number_of_bottles),
             bottle_or_bottles_after: bottle_or_bottles(number_of_bottles - 1),
             one_or_it: one_or_it(number_of_bottles) }
  end

  def verses(x, y)
    x.downto(y).collect { |bottle_num| verse(bottle_num) }.join("\n")
  end

  private
  def bottle_or_bottles(number_of_bottles)
    number_of_bottles == 1 ? "bottle" : "bottles"
  end

  def one_or_it(number_of_bottles)
    number_of_bottles == 1 ? "it" : "one"
  end

  def number_less_one(number_of_bottles)
    (number_of_bottles - 1) % 100 > 0 ? number_of_bottles - 1 : "no more"
  end
end

module BookKeeping
  VERSION = 3
end
